### PR TITLE
extract terror from pingcap/parser

### PR DESCRIPTION
**Pull request by YuJuncen**
We have extracted the error package, `terror` from `pingcap/parser`. Which can provide a more structural error type for other Go libs.

Here are some different from the `terror` package in `PingCAP/parser` to this PR:

- We removed the requirement to `pingcap/parser`, along with some methods, like `Error.ToSQLError` or `ErrClass.NewStd`.
- We use a new `Registry` instead of original some global mapping to collect created errors (see `errors.NewRegistry()`).
- We use a builder pattern like DSL to define error prototypes. 

For reviewers, the directory structure is:

- `terror_dsl`: the builder to create error prototypes.
- `terror_error`: the `Error` type, which is the prototype of a class of errors.
- `terror_gen`: the interface to auto generating the `error.toml` file.
- `terror_registry`: things about registry(i.e., `ErrClass` and `Registry`)

### CODE REVIEW

File: terror/terror.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terror
```
Comment (overvenus) [inline terror/terror.go:14]: Why not errors?

File: terror/terror.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+// https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#the-error-code-range
+type ErrCodeText string
+
+// ErrClass represents a class of errors.
+type ErrClass int
+
+var errClass2Desc = make(map[ErrClass]string)
+
+// RegisterErrorClass registers new error class for terror.
+func RegisterErrorClass(classCode int, desc string) ErrClass {
```
Comment (overvenus) [inline terror/terror.go:41]: I don't think it's a good idea to make a global registry. Different projects can define their own registry.

File: terror/terror.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+		class:   ec,
+		code:    code,
+		message: message,
+	}
+}
+
+// Error implements error interface and adds integer Class and Code, so
+// errors with different message can be compared.
+type Error struct {
+	class ErrClass
```
Comment (overvenus) [inline terror/terror.go:113]: How does this field map to RFC https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md?

File: terror/terror.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+	"go.uber.org/zap"
+)
+
+// ErrCode represents a specific error type in a error class.
+// Same error code can be used in different error classes.
+type ErrCode int
+
+// ErrCodeText is a textual error code that represents a specific error type in a error class.
+// This string should include component name, see
+// https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#the-error-code-range
```
Comment (rleungx) [inline terror/terror.go:32]: I think it's better to make sure whether we need to include the number in the ErrCodeText

File: terror/terror.go (hunk 2 above)
Comment (YuJuncen) [reply terror/terror.go:41]: But projects cannot hook at the `New` or `NewError` function. And if we want to make a generic error document generator, control of registry would be necessary.

File: terror/terror.go (hunk 4 above)
Comment (YuJuncen) [reply terror/terror.go:32]: Seems this should just be a descriptive string to this error, not the error...

**Comment by IANTHEREAL** [comment]
is this pr ready for reviewing now?

**Comment by YuJuncen** [comment]
> is this pr ready for reviewing now?

@IANTHEREAL Not yet, many refactor still needed to match the new RFC request and to make it a public library (e.g., don't use so many global variables since it might be used by many libs as @overvenus said).

**Comment by IANTHEREAL** [comment]
@YuJuncen Ok, you can determine the specific project implementation by yourself. 
If this pr is ready, you can tell guys to review it ASAP

**Comment by rleungx** [comment]
Do need an extra tool to generate the TOML file?

**Comment by YuJuncen** [comment]
> Do need an extra tool to generate the TOML file?

@rleungx at the [footer part of the RFC](https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#how-it-works), it says that we need this TOML file to build docs and something like `cargo explain`.

File: terror.go [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+// errors with different message can be compared.
+type Error struct {
+	class *ErrClass
+	code  ErrCode
+	// codeText is the textual describe of the error code
+	codeText ErrCodeText
+	// message is a template of the description of this error.
+	// printf-style formatting is enabled.
+	message string
+	// The workaround filed: how to work around this error.
```
Comment (leiysky) [inline terror.go:178]:
```suggestion
	// The workaround field: how to work around this error.
```

**Comment by YuJuncen** [comment]
> Do we need to warp this Error in parser? because there isn't `ToSQLError` here

@imtbkcat Yes, let me see... Might we can make a new struct like this:

```go
type ParserErrorExt *Error
```

Then, we can add some mix in methods at the `ParserErrorExt` type.

The risk here is that the error-defining would become dirty:

```go
var ErrSomewhat = ParserErrorExt(ClassSomewhat.DefineError().
  TextualCode("...").
  MessageTemplate("...")...)
```

Or any better idea?

**Comment by imtbkcat** [comment]
@YuJuncen  I think you can write a SQLError struct in this repo, just like:

```
type SQLError struct {
	Code    uint16
	Message string
	State   string
}
```

and this struct port the mysql/error.go in parser?

**Comment by imtbkcat** [comment]
@YuJuncen TiDB use ToSQLError just need to return an error which has `Code`, `Message` and `State` and implement `error` interface. The type will not be a matter, TiDB don't know whether it is `*mysql.SQLError`

**Comment by YuJuncen** [comment]
@imtbkcat How should I fill those fields?

**Comment by IANTHEREAL** [comment]
@rleungx @overvenus PTAL

File: terror_error.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	// It's used to teach the users how to solve the error if occurring in the real environment.
+	Workaround string
+	// Description is the expanded detail of why this error occurred.
+	// This could be written by developer at a static env,
+	// and the more detail this field explaining the better,
+	// even some guess of the cause could be included.
+	Description string
+	args        []interface{}
+	file        string
+	line        int
```
Comment (overvenus) [inline terror_error.go:64]: Can they be removed? There is no API that sets them.

File: terror_registry.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+type ErrClass struct {
+	ID          ErrClassID
+	Description string
+	errors      map[ErrorID]*Error
+	registry    *Registry
+}
+
+type ErrorID = string
+type ErrClassID = int
+type RFCErrorCode = string
```
Comment (overvenus) [inline terror_registry.go:44]: To make it clear and avoid mistakes, what about defining types?

```suggestion
type ErrorID string
type ErrClassID int
type RFCErrorCode string
```

File: terror_error.go [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+// MarshalJSON implements json.Marshaler interface.
+// aware that this function cannot save a 'registered' status,
+// since we cannot access the registry when unmarshaling,
+// and the original global registry would be removed here.
+// This function is reserved for compatibility.
+func (e *Error) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&struct {
+		Class    ErrClassID  `json:"class"`
```
Comment (overvenus) [inline terror_error.go:228]: What is class? Is it defined in RFC?

Comment (leiysky) [reply terror_error.go:228]: https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#the-error-code-range

@overvenus It's mentioned in this section.

File: terror_error.go [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+	return e.message
+}
+
+// Error implements error interface.
+func (e *Error) Error() string {
+	describe := e.codeText
+	if len(describe) == 0 {
+		describe = ErrCodeText(strconv.Itoa(int(e.code)))
+	}
+	return fmt.Sprintf("[%s:%s] %s", e.class, describe, e.getMsg())
```
Comment (ZenoTan) [inline terror_error.go:129]: I think the error should show registry, according to RFC.

**Comment by YuJuncen** [comment]
@ZenoTan @overvenus PTAL~

**Comment by ZenoTan** [comment]
> @ZenoTan @overvenus PTAL~

LGTM

File: terror_dsl.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+// MessageTemplate is the template of the error string that can be formatted after
+// calling `GenWithArgs` method.
+// currently, printf style template is used.
+func (b *Builder) MessageTemplate(template string) *Builder {
+	b.err.message = template
+	return b
+}
+
+// Done ends the define of the error.
+func (b *Builder) Done() *Error {
```
Comment (overvenus) [inline terror_dsl.go:66]: What about `Build` or `Error`?

Comment (rleungx) [reply terror_dsl.go:66]: Prefer to use `Build`

Comment (IANTHEREAL) [reply terror_dsl.go:66]: can you help to change it? then we can merge this pr

**Pull request merged**